### PR TITLE
Add git-fetch-flags

### PR DESCRIPTION
**Pull request by lox**
This adds a `git-fetch-flags` which is the missing piece for shallow clones.

### CODE REVIEW

**Comment by lox** [comment]
Pretty sure this should be enough to close https://github.com/buildkite/agent/issues/437 along with some docs. 

We could possibly support `git-depth` if it wasn't clear enough that you need `git-clone-flags=--depth=1` and `git-fetch-flags=--depth=1`.

**Comment by lox** [comment]
Would love your thoughts on a) whether this is a good idea @sj26 and b) whether `--prune` should be a default fetch param. I couldn't figure out why we didn't have `--prune` on the github pull request option.

**Comment by sj26** [comment]
Seems reasonable given our previous approach to these flags. Turning on pruning by default does also sound sensible.

I wonder if we need to teach the agent more about git and allow options like "I want shallow clones" etc? Sounds like a longer term thing though.

**Comment by lox** [comment]
@sj26 yeah, the whole point of this was to make shallow clones possible, I actually started out with a `git-clone-depth` param, and might go back there, but figured that adding `git-fetch-flags` makes it easier for folks to experiment, which seemed more inline with the philosophy of the other configuration params.

File: bootstrap/bootstrap.go [hunk 1]
```diff
@@ -1120,7 +1122,7 @@ func (b *Bootstrap) defaultCheckoutPhase() error {
 		b.shell.Commentf("Fetch and checkout pull request head from GitHub")
 		refspec := fmt.Sprintf("refs/pull/%s/head", b.PullRequest)
 
-		if err := gitFetch(b.shell, "-v", "origin", refspec); err != nil {
```
Comment (lox) [inline bootstrap/bootstrap.go]: This is the only case we don't include `--prune` @sj26, do you happen to recall why?

Comment (sj26) [reply bootstrap/bootstrap.go]: I think we just never added it 🤷‍♂️

**Pull request merged**